DOCUMENTATION.md

```diff
@@ -43,6 +41,314 @@ The overall territory is laid out here in this [C4](https://c4model.com) System
+
+<a name="why-not-queue-all-the-things"></a>
+## Processing Reactions based on a Store's Notifications vs Just using a Queue
+
+It's not uncommon for people to use a Queue (e.g. RabbitMQ, Azure Service Bus Queues, AWS SQS etc) to manage reactive processing. This typically involves wiring things up such that relevant events from the Store's notifications trigger a message on that queue. 
+
+As a system grows in features, Reaction processing tends to grow into a workflow, with complex needs regarding being able to deal with partner system outages, view the state of a workflow, cancel complex processes etc. That's normally not now it starts off though, especially in any whitepaper trying to sell you a Queue, Low Code or FaaS. 
+
+Just using a Queue has a number of pleasing properties:
+- The only thing that can prevent your Store notifications processing from getting stuck is if your Queue becomes unwriteable. Typically it'll take a lot of activity in your system for writing to the queue to become the bottleneck.
+- You have a natural way to cope with small spikes in traffic
+- You can normally easily rig things such that you run multiple instances of your Queue Processor if the backlog gets too long (i.e. the Competing Consumer pattern)
+- Anything that fails can be deferred for later retry pretty naturally (i.e. set a visibility timeout on the item)
+- Any case which for some reason triggers an exception, can be dealt with as a Poison Message
+- After a configured amount of time or attempts, you can have messages shift to a Dead Letter Queue. There are established practices regarding alerting in such circumstances and/or triggering retries of Dead Lettered messages  
+- Arranging processing via a Queue also gives a natural way for other systems (or ad-hoc admin tools) to trigger such messages too. For example, you can easily do a manual trigger or re-trigger of some reaction by adding a message out of band.
+- Queues are a generally applicable technique in a wide variety of systems, and look great on a system architecture diagram or a resume
+
+There are also a number of potential negatives; the significance of that depends on your needs:
+- You have another system to understand, worry about the uptime of, and learn to monitor
+- Marking a message invisible (or running competing consumers) will mean you can receive messages regarding a particular action out of order. In other words, you get no useful ordering guarantees (frequently this is not a problem)
+- If the time taken or processing cost of the messages on the Queue are not uniform, having 10,000 or a million messages in the queue might be hard to interpret in terms of how long it will take to clear. You also have no way to know where your high value orders are relative to that batch of 400,000 failures from that third party system 10 minute service outage that will retry in an hours time
```

Comment:
Can you elaborate on this. I think there’re some hidden assumptions.

does using store notifications guarantee uniform processing? If not, is this a problem for both the store and the queue?

For the second part I’m already concocting a scheme with a separate priority queue for my high value notifications in my mind

Reply:
I was trying to make the point that you're arguably as well off knowing you're 30000 docs behind over 2 Cosmos Physical Partitions, than to know that there are 25000 messages in the queue - if all the heavy messages failed and the easy ones are done, you'd have a false sense of progress.

Will try to word it more clearly; the intent was to say "while queues have lots of established tools and techniques, that's not a reason why it'd be intrinsically better to monitor/understand/scale things from there"

At the same time, the biggest risk I'm trying to convey is that on black Friday if some payment processor has an outage, you might find that 20% of all 4 figure sums hada validation failure. And you have a mix of items on attempt 3 and stuff in a DLQ.

If this is your real problem, having something blindly emit to a Queue immediately is no particular help
You're better off having an intentionally designed process that can do high level filtering/prioritisation in a failsafe manner first based on the store notifications
A deliverable of that might be that you want to be able to see a meaningful live query about the breakdown of important transactions in the pipeline and their state
Perhaps a queue might be part of that solution; simply replacing consuming store notifications with the same thing from a queue blidly is not where anyone should start

Also you can test store notifications stuff without a physical queue with Propulsion.MemoryStore etc

Hm, it's going to be fun to map the above ramblings to something pithy...

Reply:
OK, I wrote something. It's definitely not of the pithy variety. The detail on calculating event counts is also debatable as
- I feel it is important in the walking some folks back from the precipice re Just dumping stuff to Kafka or a Queue based on a misplaced assumption that being able to inspect them there or be able to lean on existing metrics might somehow solve the problem in a Good Enough manner
- equally, it's important to convey the diverse problems with deriving a business-meaningful metric for the various stores.
- and I want to highlight the potential for the Propulsion packages to implement some estimation logic (e.g. if I was running DynamoDb in some contexts, I'd probably implement the estimation walk logic I tease)

Ultimately this should probably become an appendix for the Reactions section (especially if the work is done to move this into a proper documentation system). 

commit: https://github.com/jet/propulsion/pull/200/commits/5eb68d49a667c09f6d72ee92b709a3c2df44b044
reading link: https://github.com/jet/propulsion/blob/doc/DOCUMENTATION.md#interpreting-ages-lags-and-gaps